### PR TITLE
publish npm packages

### LINKED ISSUE #633: Create uploader for repository type "npm"
**What would you like to be added**:
It should be possible to upload resources to an "npm" repository after transport. The uploader should use configuration from the credential provider in the OCM library to get the credentials. In addition to that configuration that contains the target npm repository is also required. Therefore a new uploader (https://github.com/open-component-model/ocm-spec/blob/main/doc/02-processing/04-digest.md?plain=1#L649) for this type is required.

It may be helpful to look into how the public Nexus repo implemented it's NPM repository to understand how to persist the blob back as npm after transport: https://github.com/sonatype/nexus-public. 

**Why is this needed**:
To be able to mimic the same repository setup on target side that is available on source side, the OCM tooling needs to provide uploaders for all existing technologies, here "npm".

### PR DESCRIPTION
**Pull request by hilmarf**
## Description

This PR fixes the feature #633.

With this we enable people to directly publish npm-blobs (aka. package.tgz containing package.json) to any registry website that implements the CommonJS Package Registry specification (e.g. artifactory).

## What type of PR is this? (check all applicable)

- [x] 🍕 Feature
- [ ] 🐛 Bug Fix
- [ ] 📝 Documentation Update
- [ ] 🎨 Style
- [ ] 🧑‍💻 Code Refactor
- [ ] 🔥 Performance Improvements
- [ ] ✅ Test
- [ ] 🤖 Build
- [ ] 🔁 CI
- [ ] 📦 Chore (Release)
- [ ] ⏩ Revert

## Related Tickets & Documents

- Closes #633

## Added tests?

- [ ] 👍 yes
- [ ] 🙅 no, because they aren't needed
- [ ] 🙋 no, because I need help
- [ ] Separate ticket for tests # (issue/pr)

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

## Added to documentation?

- [ ] 📜 README.md
- [ ] 🙅 no documentation needed

## Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: pkg/contexts/ocm/accessmethods/npm/README.md [hunk 1]
```diff
@@ -23,14 +21,12 @@ The type specific specification fields are:
 
 - **`registry`** *string*
 
-  Base URL of the NPM registry.
+  Base URL of the npmjs registry.
```
Comment (mandelsoft) [inline pkg/contexts/ocm/accessmethods/npm/README.md:24]: On the npmjs homepage they call the registry NPM REGISTRY, not npmjs registry.

File: pkg/contexts/ocm/accessmethods/npm/method.go [hunk 2]
```diff
@@ -28,11 +28,7 @@ import (
 	"github.com/open-component-model/ocm/pkg/runtime"
 )
 
-// TODO: open questions
-// - authentication???
-// - writing packages
-
-// Type is the access type of NPM registry.
+// Type is the access type of npmjs registry.
```
Comment (mandelsoft) [inline pkg/contexts/ocm/accessmethods/npm/method.go:31]: see above

File: pkg/contexts/ocm/accessmethods/npm/method.go [hunk 3]
```diff
@@ -43,7 +39,7 @@ func init() {
 	accspeccpi.RegisterAccessType(accspeccpi.NewAccessSpecType[*AccessSpec](TypeV1, accspeccpi.WithFormatSpec(formatV1), accspeccpi.WithConfigHandler(ConfigHandler())))
 }
 
-// AccessSpec describes the access for a NPM registry.
+// AccessSpec describes the access for a npmjs registry.
```
Comment (mandelsoft) [inline pkg/contexts/ocm/accessmethods/npm/method.go:42]: see above

File: pkg/contexts/ocm/accessmethods/npm/method.go [hunk 4]
```diff
@@ -57,7 +53,7 @@ type AccessSpec struct {
 
 var _ accspeccpi.AccessSpec = (*AccessSpec)(nil)
 
-// New creates a new NPM registry access spec version v1.
+// New creates a new npmjs registry access spec version v1.
```
Comment (mandelsoft) [inline pkg/contexts/ocm/accessmethods/npm/method.go:56]: see above

File: pkg/contexts/ocm/accessmethods/npm/method.go [hunk 5]
```diff
@@ -120,8 +116,6 @@ func (a *AccessSpec) getPackageMeta(ctx accspeccpi.Context) (*meta, error) {
 	return &metadata, nil
 }
 
-////////////////////////////////////////////////////////////////////////////////
-
```
Comment (mandelsoft) [inline pkg/contexts/ocm/accessmethods/npm/method.go:124]: this comment separates two object types, spec and method and should be kept.

File: pkg/contexts/ocm/blobhandler/handlers/generic/npmjs/blobhandler.go [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+	}
+	defer blobReader.Close()
+
+	data, err := io.ReadAll(blobReader)
+	if err != nil {
+		return nil, err
+	}
+
+	// read package.json from tarball to get name, version, etc.
+	log := logging.Context().Logger()
```
Comment (mandelsoft) [inline pkg/contexts/ocm/blobhandler/handlers/generic/npmjs/blobhandler.go:48]: you should introduce an own realm for mpm and use a logger for that realm.

File: pkg/contexts/ocm/blobhandler/handlers/generic/npmjs/blobhandler.go [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+	log.Debug("reading package.json from tarball")
+	var pkg *Package
+	pkg, err = prepare(data)
+	if err != nil {
+		return nil, err
+	}
+	tbName := pkg.Name + "-" + pkg.Version + ".tgz"
+	pkg.Dist.Tarball = b.spec.Url + pkg.Name + "/-/" + tbName
+	log.Debug("identified package: %s@%s", pkg.Name, pkg.Version)
```
Comment (mandelsoft) [inline pkg/contexts/ocm/blobhandler/handlers/generic/npmjs/blobhandler.go:57]: Use key/value pairs in arguments.
If you want a formatted message use {{<key>}} to refer to the value of a dedicated key.

File: pkg/contexts/ocm/blobhandler/handlers/generic/npmjs/blobhandler.go [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+
+	// use user+pass+mail from credentials to login and retrieve bearer token
+	cred := GetCredentials(ctx.GetContext(), b.spec.Url, pkg.Name)
+	username := cred[ATTR_USERNAME]
+	password := cred[ATTR_PASSWORD]
+	email := cred[ATTR_EMAIL]
+	if username == "" || password == "" || email == "" {
+		return nil, fmt.Errorf("username, password or email missing")
+	}
+	log.Debug("logging in as %s on %s", username, b.spec.Url)
```
Comment (mandelsoft) [inline pkg/contexts/ocm/blobhandler/handlers/generic/npmjs/blobhandler.go:67]: see log format

File: pkg/contexts/ocm/blobhandler/handlers/generic/npmjs/blobhandler.go [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+		return nil, err
+	}
+
+	// check if package exists
+	exists, err := packageExists(b.spec.Url, *pkg, token)
+	if err != nil {
+		return nil, err
+	}
+	if exists {
+		log.Debug("package %s@%s already exists, skipping upload", pkg.Name, pkg.Version)
```
Comment (mandelsoft) [inline pkg/contexts/ocm/blobhandler/handlers/generic/npmjs/blobhandler.go:79]: see log format

File: pkg/contexts/ocm/blobhandler/handlers/generic/npmjs/blobhandler.go [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+	req, err := http.NewRequestWithContext(context.Background(), http.MethodPut, b.spec.Url+"/"+url.PathEscape(pkg.Name), bytes.NewReader(marshal))
+	if err != nil {
+		return nil, err
+	}
+	req.Header.Set("authorization", "Bearer "+token)
+	req.Header.Set("content-type", "application/json")
+
+	// send PUT request - upload tgz
+	client := http.Client{}
+	log.Debug("uploading package %s@%s to %s", pkg.Name, pkg.Version, b.spec.Url)
```
Comment (mandelsoft) [inline pkg/contexts/ocm/blobhandler/handlers/generic/npmjs/blobhandler.go:112]: see log format

File: pkg/contexts/ocm/blobhandler/handlers/generic/npmjs/blobhandler.go [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode != http.StatusCreated {
+		all, err := io.ReadAll(resp.Body)
+		if err != nil {
+			return nil, err
+		}
+		return nil, fmt.Errorf("http (%d) - failed to upload package: %s", resp.StatusCode, string(all))
+	}
+	log.Debug("package %s@%s uploaded to %s", pkg.Name, pkg.Version, b.spec.Url)
```
Comment (mandelsoft) [inline pkg/contexts/ocm/blobhandler/handlers/generic/npmjs/blobhandler.go:125]: see log format

File: pkg/contexts/ocm/blobhandler/handlers/generic/npmjs/const.go [hunk 12]
```diff
@@ -0,0 +1,16 @@
+package npmjs
```
Comment (mandelsoft) [inline pkg/contexts/ocm/blobhandler/handlers/generic/npmjs/const.go:1]: I would name it just `npm`

File: pkg/contexts/ocm/blobhandler/handlers/generic/npmjs/identity.go [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+)
+
+func init() {
+	attrs := listformat.FormatListElements("", listformat.StringElementDescriptionList{
+		ATTR_USERNAME, "the basic auth user name",
+		ATTR_PASSWORD, "the basic auth password",
+		ATTR_EMAIL, "npmjs registries, require an email address",
+	})
+
+	cpi.RegisterStandardIdentity(CONSUMER_TYPE, hostpath.IdentityMatcher(CONSUMER_TYPE), `Npmjs repository
```
Comment (mandelsoft) [inline pkg/contexts/ocm/blobhandler/handlers/generic/npmjs/identity.go:21]: NPM

File: pkg/contexts/ocm/blobhandler/handlers/generic/npmjs/publish.go [hunk 14]
```diff
@@ -0,0 +1,214 @@
...
+	if resp.StatusCode != http.StatusOK {
+		return false, fmt.Errorf("http (%d) - %s", resp.StatusCode, string(all))
+	}
+	var data map[string]interface{}
+	err = json.Unmarshal(all, &data)
+	if err != nil {
+		return false, err
+	}
+	dist := data["dist"].(map[string]interface{})
+	if pkg.Dist.Integrity == dist["integrity"] {
```
Comment (mandelsoft) [inline pkg/contexts/ocm/blobhandler/handlers/generic/npmjs/publish.go:204]: Is it always the same digest type, or should we check the type analysing the digest prefix (<type>:<hash>)

File: pkg/contexts/ocm/blobhandler/handlers/generic/npmjs/registration.go [hunk 15]
```diff
@@ -0,0 +1,76 @@
...
+
+	return true, nil
+}
+
+func (r *RegistrationHandler) GetHandlers(ctx cpi.Context) registrations.HandlerInfos {
+	return registrations.NewLeafHandlerInfo("uploading npmjs artifacts", `
+The <code>npmjsArtifacts</code> uploader is able to upload npmjs artifacts
+as artifact archive according to the npmjs package spec.
+The following artifact media types are supported: `+mime.MIME_TGZ+`
+By default, it is registered for these mimetypes.
```
Comment (mandelsoft) [inline pkg/contexts/ocm/blobhandler/handlers/generic/npmjs/registration.go:69]: There is no default registration. I think you mean: If registered the default mime type is ...

File: pkg/contexts/ocm/blobhandler/handlers/generic/npmjs/registration.go [hunk 16]
```diff
@@ -0,0 +1,76 @@
...
+
+It accepts a config with the following fields:
+'url': the URL of the npmjs repository.
+If not given, the default npmjs.com repository.
```
Comment (mandelsoft) [inline pkg/contexts/ocm/blobhandler/handlers/generic/npmjs/registration.go:73]: or a plain string for the URL

**Review by mandelsoft — CHANGES_REQUESTED** [review]

File: pkg/contexts/ocm/blobhandler/handlers/generic/npmjs/publish.go (hunk 14 above)
Comment (hilmarf) [reply pkg/contexts/ocm/blobhandler/handlers/generic/npmjs/publish.go:204]: as far as I've seen it, it's always the same
https://docs.npmjs.com/cli/v6/configuring-npm/package-lock-json#integrity

when SHA512 isn't sufficient anymore for NPM, then we can adapt the coding

File: pkg/contexts/ocm/accessmethods/npm/method.go (hunk 5 above)
Comment (hilmarf) [reply pkg/contexts/ocm/accessmethods/npm/method.go]: aha! ok

File: pkg/contexts/ocm/blobhandler/handlers/generic/npmjs/registration.go (hunk 15 above)
Comment (hilmarf) [reply pkg/contexts/ocm/blobhandler/handlers/generic/npmjs/registration.go]: re-written

**Review by mandelsoft — APPROVED** [review]

**Pull request merged**